### ISSUE TITLE
Document the convention that pixel[0,0] centre lies at (0.5, 0.5)

### ISSUE THREAD

**Issue by stefanv** [scikit-image/scikit-image#1732]
And describe the implications this has for boundary extension and interpolation.

**Comment by sumitbinnani** [comment]
Hi @stefanv, are you talking about something like:

https://msdn.microsoft.com/en-us/library/windows/desktop/bb219690%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396?

If the input image image being processed is already digitized, it should not have implications on boundary extension or interpolation? Kindly correct me if I am not able to understand your problem statement.

**Comment by stefanv** [comment]
Yes, except @ahojnnes indicated that the centers are at (0.5, 0.5) in our case.  Where you choose the pixel origin has implications on how interpolation and boundary extension is handled when it comes to nearest neighbor interpolation, e.g.

**Comment by jni** [comment]
@stefanv I'm only just coming to terms with [pixels not being squares](http://alvyray.com/Memos/CG/Microsoft/6_pixel.pdf), but it seems to me that "centers at .5" is inconsistent with this view. Am I wrong?

Example: if we want to sample rom `img` at (1., 1.), do we really want to interpolate between all the points in `img[:2, :2]`, or do we want to just return `img[1, 1]`?

At the very least the document would clarify these things in my head... =)

**Comment by stefanv** [comment]
No, you are not wrong; you are highlighting the concern I raised with @ahojnnes.  But I think we all agree that it is more-or-less fine whatever we choose, as long as we make a nice diagram and include it in the docs.

Currently, we interpolate to get `img[1, 1]`.  If I had to redesign everything from scratch, I might have gone for `img[1,1]` being the actual value at `img[1,1]`.

**Comment by gazzar** [comment]
@stefanv I'm a bit confused by this. By default the rotation centre of the rotate() transform is taken about center=(rows / 2 - 0.5, cols / 2 - 0.5). Shouldn't this be about center=(rows / 2, cols / 2) instead? I see that test_rotate() in test_warps.py seems to test for this though.

**Comment by leducvin** [comment]
@gazzar I don't know if this is relevant really, but `matplotlib.pyplot.imshow` indeed shows the center of an array to be at `(rows / 2 - 0.5, cols / 2 - 0.5)`. Looking at the source code of the `skimage.transform.rotate` function, `rows` is interpreted as `image.shape[0]` and `cols` as `image.shape[1]`.

[Actually, `center = np.array((cols, rows)) / 2. - 0.5` is used in the source code and not `center = np.array((rows, cols)) / 2. - 0.5.`, I have created issue #2671 for this, referenced above by @soupault]

This implies that, for matplotlib at least, the center of the pixel at (0, 0) lies at (0, 0). The pixel at (0, 0) is _rendered as spanning a region from -0.5 to 0.5 in both axes_, meaning the top left-corner of the displayed image is at (-0.5, -0.5).

In my testing of the rotate function, if I use `(rows / 2, cols / 2)`, it does not seem to behave correctly. Therefore, **the rotate function seems to be following matplotlib's convention**.

skimage documentation says:
> Two-dimensional (2D) grayscale images (such as camera above) are indexed by row and columns (abbreviated to either (row, col) or (r, c)), with the lowest element (0, 0) at the top-left corner.

...which clearly states the top-left corner pixel is at (0, 0), but it is rather vague if this means the center is modeled as being at (0.5, 0.5) or modeled as being at (0, 0). It seems to me to most logical way is to consider pixel values (which are point values, as @jni mentions above) to sit exactly at their index-coordinates.

So yes, this is really confusing.

Here is a little demonstration:
```
import matplotlib.pyplot as plt
import numpy
import skimage.transform

# Image of shape (9, 9)
# img[4, 4] is the center element
img = numpy.array(
    [
        [0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 1, 1, 1, 0, 0, 0],
        [0, 0, 2, 1, 0, 1, 0, 0, 0],
        [0, 0, 0, 1, 1, 1, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0],
    ]
)

fig = plt.figure()
plt.imshow(img)
plt.title('Original image')

img_rot_center = skimage.transform.rotate(
    img, 90, order=0, preserve_range=True)

fig = plt.figure()
plt.imshow(img_rot_center)
plt.title('Image rotated using center=None')

# Try to rotate around the center manually, by using
# center = (rows / 2 - 0.5, cols / 2 - 0.5)
img_rot_center_manual_1 = skimage.transform.rotate(
    img, 90, order=0, center=(4, 4), preserve_range=True)

fig = plt.figure()
plt.imshow(img_rot_center_manual_1)
plt.title('Image rotated using center=(rows / 2 - 0.5, cols / 2 - 0.5)')

# Try to rotate around the center manually, by using
# center = (rows / 2, cols / 2)
img_rot_center_manual_2 = skimage.transform.rotate(
    img, 90, order=0, center=(4.5, 4.5), preserve_range=True)

fig = plt.figure()
plt.imshow(img_rot_center_manual_2)
plt.title('Image rotated using center=(rows / 2, cols / 2)')
plt.show()

# This works fine
assert img[4, 4] == img_rot_center[4, 4]

# This works fine
assert img[4, 4] == img_rot_center_manual_1[4, 4]
assert numpy.allclose(img_rot_center, img_rot_center_manual_1)

# This fails
assert img[4, 4] == img_rot_center_manual_2[4, 4]
assert numpy.allclose(img_rot_center, img_rot_center_manual_2)
```

Please tell me if I've gone wrong somewhere.

@stefanv What do you mean? I don't understand.
> Currently, we interpolate to get img[1, 1]. If I had to redesign everything from scratch, I might have gone for img[1,1] being the actual value at img[1,1]

If I ask img[4, 4] in the numpy array in the code above, it most definitely returns 0, no interpolation needed. Similarly, skimage docs give this as an example:
```
>>> from skimage import data
>>> camera = data.camera()
>>> type(camera)
<type 'numpy.ndarray'>
# Get the value of the pixel on the 10th row and 20th column
>>> camera[10, 20]
153
```
Again, there doesn't seem to be any interpolation involved. Is `img` a `numpy.ndarray` in the comment you posted?

**Comment by jni** [comment]
Just a note here: as we deprecate transposed (col/row) coordinates in `transform`, we certainly have an opportunity to center coordinates on the pixels. I will argue very strongly that we should do this. =) This would also make us consistent with `ndimage`, which is a major plus.

**Comment by JDWarner** [comment]
I agree with @jni, a point sample model where the datum is associated with the exact specified location would be the preferred model - from the DSP theory standpoint in addition to agreement with SciPy.

**Comment by stefanv** [comment]
@grlee77 You have worked a fair bit on interpolation; would you be willing to take a look at this documentation issue?
